File: python/day21/part2.py
#!/usr/bin/env python3
import time
import part1


def solve_part2(tree):
    first = part1.compute(tree.left)
    second = part1.compute(tree.right)
    return round((first.real - second.real) / (second.imag - first.imag))


if __name__ == "__main__":
    start = time.time()
    parsed = part1.parse("data.txt", True)
    result = solve_part2(parsed)
    print(f'Part 2 : {result}  ({round(1000 * (time.time() - start))} ms)')

File: c++/day21/part2.cpp
#include "part2.h"

using namespace std;

namespace Day21 {
    namespace Part2 {

        void updateMonkey(
                const std::shared_ptr<Monkey<complex<long double>>>& monkey,
                const string& name,
                const complex<long double> val) {
            if (monkey != nullptr) {
                if (monkey->name == name) {
                    monkey->val = val;
                }
                updateMonkey(monkey->left, name, val);
                updateMonkey(monkey->right, name, val);
            }
        }

        long long solve(const std::shared_ptr<Monkey<complex<long double>>>& rootMonkey) {
            updateMonkey(rootMonkey, "humn", complex<long double>{ 0, 1 });
            const auto left  = rootMonkey->left->getVal();
            const auto right = rootMonkey->right->getVal();
            // the value shouted by the "humn" monkey is on the left side
            return static_cast<long long>(round((right.real() - left.real()) / left.imag()));
        }
    }
}
